### PR TITLE
libusb backend: stabilize device path, superseeds: #117

### PR DESCRIPTION
**Pull request by bearsh**
format the path like the linux kernel does:
<busnum>-<port[.port]>…:<config num>.<interface num>

See also https://github.com/libusb/hidapi/pull/117

Closes: #117

### CODE REVIEW

**Review by Youw — APPROVED** [review]
This looks perfect.
I'll test this out over the weekend, but I think we should be good with currently suggested implementation.

**Comment by derekatkins** [comment]
Looks good to me.

**Review by derekatkins — APPROVED** [review]

File: libusb/hid.c [hunk 1]
```diff
@@ -484,15 +484,29 @@ static wchar_t *get_usb_string(libusb_device_handle *dev, uint8_t idx)
...
+static char *make_path(libusb_device *dev, int interface_number, int config_number)
 {
-	char str[64];
-	snprintf(str, sizeof(str), "%04x:%04x:%02x",
-		libusb_get_bus_number(dev),
-		libusb_get_device_address(dev),
-		interface_number);
-	str[sizeof(str)-1] = '\0';
-
+	char str[64]; // max length "000-000.000.000.000.000.000.000:000.000"
```
Comment (Youw) [inline libusb/hid.c:489]:
```suggestion
	char str[64]; /* max length "000-000.000.000.000.000.000.000:000.000" */
```

File: libusb/hid.c [hunk 2]
```diff
@@ -484,15 +484,29 @@ static wchar_t *get_usb_string(libusb_device_handle *dev, uint8_t idx)
...
+	// Note that USB3 port count limit is 7; use 8 here for alignment
```
Comment (Youw) [inline libusb/hid.c:490]:
```suggestion
	/* Note that USB3 port count limit is 7; use 8 here for alignment */
```

File: libusb/hid.c [hunk 3]
```diff
@@ -484,15 +484,29 @@ static wchar_t *get_usb_string(libusb_device_handle *dev, uint8_t idx)
...
+
+	if (num_ports > 0) {
+		int n = snprintf(str, sizeof("000-000"), "%u-%u", libusb_get_bus_number(dev), port_numbers[0]);
+		for (uint8_t i = 1; i < num_ports; i++) {
+			n += snprintf(&str[n], sizeof(".000"), ".%u", port_numbers[i]);
+		}
+		n += snprintf(&str[n], sizeof(":000.000"), ":%u.%u", (uint8_t)config_number, (uint8_t)interface_number);
+		str[n] = '\0';
+	} else {
+		// USB3.0 specs limit number of ports to 7 and buffer size here is 8
```
Comment (Youw) [inline libusb/hid.c:502]:
```suggestion
		/* USB3.0 specs limit number of ports to 7 and buffer size here is 8 */
```

**Review by Youw — COMMENTED** [review]
By agreement we're using "C-style" comments only in this source file

**Comment by bearsh** [comment]
> By agreement we're using "C-style" comments only in this source file

`//` is AFAIK C99 and above... anyway, pull request has been updated

**Comment by Youw** [comment]
> `//` is AFAIK C99 and above... 

Yeah. It is more about agreement/code-style/source consistency.

**Comment by Youw** [comment]
I've tested this out with several of my devices. With and without hubs (multiple hubs) - so far looking good!

**Comment by netfab** [comment]
Tested successfully with my keyboard driver.

> - 09:04:26.441                    openUSBDevice  - searched path: 3-2:1.1
> - 09:04:26.504                    openUSBDevice  - detected HIDAPI device path: 3-2:1.0
> - 09:04:26.504                    openUSBDevice - detected HIDAPI device path: 3-2:1.1
> - 09:04:26.504                    openUSBDevice - [b3d2] found HIDAPI device: 3-2:1.1
> - 09:04:26.520                    openUSBDevice - [b3d2] opened HIDAPI USB device
> - 09:04:26.520      sendUSBDeviceInitialization - [b3d2] sending Logitech G510s Gaming Keyboard initialization requests
> - 09:04:26.525               sendControlRequest - sent HIDAPI feature report: 19 bytes - expected: 19
> - 09:04:26.528               sendControlRequest - sent HIDAPI feature report: 8 bytes - expected: 8

**Pull request merged**